cmd/k8s-operator/tsrecorder.go

```diff
@@ -0,0 +1,365 @@
+		s.Spec = ss.Spec
+	}); err != nil {
+		return fmt.Errorf("error creating stateful set: %w", err)
+	}
+
+	var devices []tsapi.TailnetDevice
+
+	tsHost, ips, ok, err := r.getDeviceInfo(ctx, tsr.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get device info: %w", err)
+	}
+	if !ok {
+		logger.Debugf("no Tailscale hostname known yet, waiting for tsrecorder pod to finish auth")
+		return nil
+	}
+
+	devices = append(devices, tsapi.TailnetDevice{
+		Hostname:   tsHost,
+		TailnetIPs: ips,
+	})
+
+	tsr.Status.Devices = devices
+
+	return nil
+}
+
+// maybeCleanup just deletes the device from the tailnet. All the kubernetes
+// resources linked to a TSRecorder will get cleaned up via owner references
+// (which we can use because they are all in the same namespace).
+func (r *TSRecorderReconciler) maybeCleanup(ctx context.Context, tsr *tsapi.TSRecorder) (bool, error) {
+	logger := r.logger(tsr.Name)
+
+	id, ok, err := r.getNodeID(ctx, tsr.Name)
+	if err != nil {
+		return false, err
+	}
+	if !ok {
+		logger.Debugf("secret %s-0 not found or does not contain node ID, continuing cleanup", tsr.Name)
+		r.mu.Lock()
+		r.tsRecorders.Remove(tsr.UID)
+		gaugeTSRecorderResources.Set(int64(r.tsRecorders.Len()))
+		r.mu.Unlock()
+		return true, nil
+	}
+
+	logger.Debugf("deleting device %s from control", string(id))
+	if err := r.tsClient.DeleteDevice(ctx, string(id)); err != nil {
+		errResp := &tailscale.ErrResponse{}
+		if ok := errors.As(err, errResp); ok && errResp.Status == http.StatusNotFound {
+			logger.Debugf("device %s not found, likely because it has already been deleted from control", string(id))
+		} else {
+			return false, fmt.Errorf("error deleting device: %w", err)
+		}
+	} else {
+		logger.Debugf("device %s deleted from control", string(id))
+	}
+
+	// Unlike most log entries in the reconcile loop, this will get printed
+	// exactly once at the very end of cleanup, because the final step of
+	// cleanup removes the tailscale finalizer, which will make all future
+	// reconciles exit early.
+	logger.Infof("cleaned up TSRecorder resources")
+	r.mu.Lock()
+	r.tsRecorders.Remove(tsr.UID)
+	gaugeTSRecorderResources.Set(int64(r.tsRecorders.Len()))
+	r.mu.Unlock()
+	return true, nil
+}
+
+func (r *TSRecorderReconciler) ensureAuthSecretCreated(ctx context.Context, tsr *tsapi.TSRecorder) error {
+	logger := r.logger(tsr.Name)
+	key := types.NamespacedName{
+		Namespace: r.tsNamespace,
+		Name:      tsr.Name,
+	}
+	if err := r.Get(ctx, key, &corev1.Secret{}); err == nil {
+		// No updates, already created the auth key.
+		logger.Debugf("secret %s/%s already exists", key.Namespace, key.Name)
+		return nil
+	} else if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Create API Key secret which is going to be used by the statefulset
+	// to authenticate with Tailscale.
+	logger.Debugf("creating authkey for new tsrecorder")
+	tags := tsr.Spec.Tags
+	if len(tags) == 0 {
+		tags = tsapi.Tags{"tag:k8s-recorder"}
```

Comment:
My main thought was that users will probably most frequently specify recorders (in ACLs) via tag, and so they'll most commonly need a different tag from any other k8s resources deployed. However, agreed that it would be nice to maintain the simple onboarding, and deploying recorders is a much more niche use case, so I'll switch to the `k8s` tag.